hr_contract_cae/models/hr_contract.py

```diff
@@ -19,20 +144,162 @@ class Contract(models.Model):
+    attachment_ids = fields.One2many(
+        comodel_name="ir.attachment",
+        inverse_name="res_id",
+        string="Attachments",
+        domain=[("res_model", "=", "hc.contract")],
+    )
+
+    @api.model
+    def create(self, vals):
+        res = super().create(vals)
+        if not res.contract_group_id:
+            contract_group = self.env["hr.contract.group"].create({})
+            res.contract_group_id = contract_group
+        if res.amendment_index == 0:
+            res.initial_contract_id = res.get_initial()
+        return res
+
+    @api.multi
+    @api.depends("employee_id", "initial_contract_id", "type_id")
+    def _compute_name(self):
+        for contract in self:
+            contract.name = "/".join(
+                filter(
+                    None,
+                    [
+                        (contract.employee_id.name or ""),
+                        (
+                            contract.initial_contract_id.type_id.name
+                            if contract.amendment_index > 0
+                            else ""
+                        ),
+                        (
+                            str(contract.amendment_index)
+                            if contract.amendment_index > 0
+                            else ""
+                        ),
+                        (contract.type_id.name or ""),
+                    ],
+                )
+            )
+
+    @api.multi
+    @api.depends("amendment_index")
+    def _compute_type_echelon(self):
+        for contract in self:
+            contract.type_echelon = (
+                "main" if contract.amendment_index == 0 else "amendment"
+            )
 
     @api.multi
     @api.depends("hours", "hourly_wage")
     def _compute_wage(self):
         for contract in self:
-            contract.wage = self.hours * self.hourly_wage
+            contract.wage = contract.hours * contract.hourly_wage
 
     @api.onchange("date_start", "duration")
     def onchange_date_start_duration(self):
+        if not self.duration:
+            self.date_end = False
         if self.date_start and self.duration:
-            self.date_end = self.date_start + relativedelta(months=self.duration)
+            self.date_end = self.date_start + relativedelta(
+                months=self.duration
+            )
 
     @api.onchange("date_end")
     def onchange_date_end(self):
+        if not self.date_end:
+            self.duration = False
         if self.date_start and self.date_end:
             rd = relativedelta(self.date_end, self.date_start)
             self.duration = rd.months + rd.years * 12
+
+    @api.multi
+    @api.depends("contract_group_contract_ids", "type_id")
+    def _compute_type_count(self):
+        for contract in self:
+            type_count = 0
+            for c in contract.contract_group_contract_ids:
+                if c.type_id == contract.type_id:
+                    type_count += 1
+            contract.type_count = type_count
```

Comment:
This feels like it should be on the group. Plus, now that you don't have to traverse the contract chain, you could do:

```python
contract.type_count = len(list(filter(
    λ c: c.type_id == contract.type_id,
    contract.contract_group_contract_ids)))
# or
contract.type_count = len([c for c in contract.contract_group_contract_ids
                           if c.type_id == contract.type_id])
```